### PR TITLE
Druntime part of fix for 18559

### PR DESCRIPTION
**Pull request by thewilsonator**
druntime part https://github.com/dlang/dmd/pull/7995

### CODE REVIEW

File: src/core/math.d [hunk 1]
```diff
@@ -40,7 +49,9 @@ public:
  *      Results are undefined if |x| >= $(POWER 2,64).
  */
 
-real cos(real x) @safe pure nothrow;       /* intrinsic */
+float cos(float x);     /* intrinsic */
```
Comment (JackStouffer) [inline src/core/math.d]: Shouldn't this be `cosf` and the `real` be `cosl`? Same throughout.

Comment (thewilsonator) [reply src/core/math.d]: intrinsics are detected based on the mangled name https://github.com/dlang/dmd/pull/7995/files#diff-34b7829e9285bc4fe4e718f0653a6115R421
so no.

File: src/core/math.d [hunk 2]
```diff
@@ -145,15 +153,22 @@ real fabs(real x) @safe pure nothrow;      /* intrinsic */
...
  * Building block functions, they
  * translate to a single x87 instruction.
  */
-
-real yl2x(real x, real y)   @safe pure nothrow;       // y * log2(x)
-real yl2xp1(real x, real y) @safe pure nothrow;       // y * log2(x + 1)
+// y * log2(x)
+float yl2x(float x);    /* intrinsic */
+double yl2x(double x);  /* intrinsic */ /// ditto
+real yl2x(real x);      /* intrinsic */ /// ditto
```
Comment (n8sh) [inline src/core/math.d]: I think `yl2x` should have 2 arguments.

Comment (thewilsonator) [reply src/core/math.d]: True.

**Comment by quickfur** [comment]
LGTM.

**Comment by JinShil** [comment]
@thewilsonator Are you still interested in pursuing this?  It appears there are `float`/`real` mismatches.

**Comment by thewilsonator** [comment]
Yes, what mismatches are there (I've just got back from Europe so I'm not thinking properly yet)?

**Comment by JinShil** [comment]
> what mismatches are there

See the output of the test suite.

**Comment by MoonlightSentinel** [comment]
What's the status of this PR? The dmd PR has been merged already...

**Comment by thewilsonator** [comment]
Just rebased, lets see what happens.

File: src/core/math.d [hunk 3]
```diff
@@ -177,38 +192,29 @@ unittest
...
 T toPrec(T:real)(real f)  { pragma(inline, false); return f; }
 
 @safe unittest
 {
-    static float f = 1.1f;
-    static double d = 1.1;
-    static real r = 1.1L;
+    float f = 1.1f;
+    double d = 1.1;
+    real r = 1.1L;
```
Comment (thewilsonator) [inline src/core/math.d:217]: @ibuclaw was there any reason these need to be static? This makes the unittest import.

File: src/core/math.d [hunk 4]
```diff
@@ -28,6 +28,15 @@ module core.math;
...
 
+/*****************************************
+ * Returns x rounded to a long value using the FE_TONEAREST rounding mode.
+ * If the integer value of x is
+ * greater than long.max, the result is
+ * indeterminate.
+ */
+extern (C) real rndtonl(real x);
+
+@safe: pure: nothrow:
```
Comment (n8sh) [inline src/core/math.d:39]: What is the reason that `rndtonl` is not `@safe` and `nothrow`?

**Comment by thewilsonator** [comment]
Needs https://github.com/dlang/phobos/pull/7461

File: src/core/math.d (hunk 3 above)
Comment (ibuclaw) [reply src/core/math.d:219]: To force the code generator (not CTFE) to do the work.  Practically the same is done in the test suite, so could remove if you really must.

Comment (ibuclaw) [reply src/core/math.d:219]: I'll remove it as part of #2871

File: src/core/math.d (hunk 4 above)
Comment (thewilsonator) [reply src/core/math.d]: Fixed.

File: src/core/math.d (hunk 3 above)
Comment (thewilsonator) [reply src/core/math.d:219]: That PR seems to have [offended the gods of continuous integration](https://github.com/dlang/druntime/pull/2871#issuecomment-612826938), so I'd rather that be done here.

**Comment by thewilsonator** [comment]
> Needs dlang/phobos#7461

Is now merged, this should be good to go.

File: src/core/math.d (hunk 3 above)
Comment (ibuclaw) [reply src/core/math.d:219]: I've added approxEqual, so the Win32 targets should now be happy with their inaccuracies.

**Comment by thewilsonator** [comment]
Ping @ibuclaw @n8sh

**Pull request merged**